### PR TITLE
[feat] Support partitioned topic reader.

### PR DESCRIPTION
**Pull request by shibd**
### Motivation

In the current implementation, Reader does not yet support partitioned topics.

### Modifications

- MultiTopicConsumer support `hasMessageAvailable` and `acknowledgeCumulativeAsync` method.
- Reader support partitioned topic.

### Verifying this change
- Add all tests in `ReaderTest` to multi-partition topic scenarios.
- Add `ConsumerTest.testAcknowledgeCumulativeWithPartition` unit test to cover MultiTopicConsumer.acknowledgeCumulativeAsync method.

### Documentation

- [ ] `doc-required` 
(Your PR needs to update docs and you will update later)

- [x] `doc-not-needed` 
(Please explain why)

- [ ] `doc` 
(Your PR contains doc changes)

- [ ] `doc-complete`
(Docs have been already added)

### CODE REVIEW

File: include/pulsar/ConsumerConfiguration.h [hunk 1]
```diff
@@ -383,6 +383,26 @@ class PULSAR_PUBLIC ConsumerConfiguration {
...
+     *
+     * <p>Options are:
+     * <ul>
+     *  <li>{@link SubscriptionMode#Durable} (Default)</li>
+     *  <li>{@link SubscriptionMode#NonDurable}</li>
+     * </ul>
+     *
+     * @param subscriptionMode the subscription mode value
+     */
+    void setSubscriptionMode(SubscriptionMode subscriptionMode);
```
Comment (BewareMyPower) [inline include/pulsar/ConsumerConfiguration.h:397]:
```suggestion
    ConsumerConfiguration& setSubscriptionMode(SubscriptionMode subscriptionMode);
```

Support method chaining.

Comment (shibd) [reply include/pulsar/ConsumerConfiguration.h:397]: I noticed that the style was different. Some returned config&, and some did not. I will unify it.

File: include/pulsar/ConsumerConfiguration.h [hunk 2]
```diff
@@ -383,6 +383,26 @@ class PULSAR_PUBLIC ConsumerConfiguration {
...
+     * @param subscriptionMode the subscription mode value
+     */
+    ConsumerConfiguration& setSubscriptionMode(SubscriptionMode subscriptionMode);
+
+    /**
+     * Get subscription mode.
+     *
+     * @return
+     */
+    SubscriptionMode getSubscriptionMode() const;
```
Comment (BewareMyPower) [inline include/pulsar/ConsumerConfiguration.h:404]: It's better not support setting the subscription mode. /cc @codelipenghui

**Review by BewareMyPower — CHANGES_REQUESTED** [review]

**Comment by BewareMyPower** [comment]
@shibd Could you resolve the conflicts and address the comments?

File: include/pulsar/ConsumerConfiguration.h (hunk 2 above)
Comment (shibd) [reply include/pulsar/ConsumerConfiguration.h]: done.

File: lib/ClientConnection.h [hunk 3]
```diff
@@ -65,6 +65,10 @@ class ConsumerImpl;
 typedef std::shared_ptr<ConsumerImpl> ConsumerImplPtr;
 typedef std::weak_ptr<ConsumerImpl> ConsumerImplWeakPtr;
 
+class ConsumerImplBase;
+typedef std::shared_ptr<ConsumerImplBase> ConsumerImplBasePtr;
+typedef std::weak_ptr<ConsumerImplBase> ConsumerImplBaseWeakPtr;
```
Comment (BewareMyPower) [inline lib/ClientConnection.h:70]: These forward declarations are meaningless.

File: lib/ReaderImpl.h [hunk 4]
```diff
@@ -58,8 +58,9 @@ extern PULSAR_PUBLIC ConsumerConfiguration consumerConfigOfReader;
 
 class PULSAR_PUBLIC ReaderImpl : public std::enable_shared_from_this<ReaderImpl> {
    public:
-    ReaderImpl(const ClientImplPtr client, const std::string& topic, const ReaderConfiguration& conf,
-               const ExecutorServicePtr listenerExecutor, ReaderCallback readerCreatedCallback);
+    ReaderImpl(const ClientImplPtr client, const std::string& topic, const int partitions,
```
Comment (BewareMyPower) [inline lib/ReaderImpl.h:61]: Don't use a const value as a parameter. See a similar comment here: https://github.com/apache/pulsar-client-cpp/pull/169#discussion_r1065684370

File: tests/ReaderTest.cc [hunk 5]
```diff
@@ -36,10 +36,69 @@ using namespace pulsar;
 static std::string serviceUrl = "pulsar://localhost:6650";
 static const std::string adminUrl = "http://localhost:8080/";
 
-TEST(ReaderTest, testSimpleReader) {
+TEST(ReaderTest, testPartitionIndex) {
```
Comment (BewareMyPower) [inline tests/ReaderTest.cc:39]: Why did you move the `testPartitionIndex` here? Please avoid making these unrelated changes.

File: tests/ReaderTest.cc [hunk 6]
```diff
@@ -606,3 +653,5 @@ TEST(ReaderTest, testReceiveAfterSeek) {
 
     client.close();
 }
+
+INSTANTIATE_TEST_CASE_P(Pulsar, ReaderTest, ::testing::Values(true, false));
```
Comment (BewareMyPower) [inline tests/ReaderTest.cc:657]:
```suggestion
INSTANTIATE_TEST_SUITE_P(Pulsar, ReaderTest, ::testing::Values(true, false));
```

`INSTANTIATE_TEST_CASE_P` is deprecated.

**Comment by shibd** [comment]
@BewareMyPower @RobertIndie PTAL, Because it blocked TableView pr.

File: lib/MultiTopicsConsumerImpl.h [hunk 7]
```diff
@@ -53,10 +54,15 @@ class MultiTopicsConsumerImpl : public ConsumerImplBase {
    public:
     MultiTopicsConsumerImpl(ClientImplPtr client, TopicNamePtr topicName, int numPartitions,
                             const std::string& subscriptionName, const ConsumerConfiguration& conf,
-                            LookupServicePtr lookupServicePtr);
+                            LookupServicePtr lookupServicePtr,
+                            const Commands::SubscriptionMode = Commands::SubscriptionModeDurable,
```
Comment (BewareMyPower) [inline lib/MultiTopicsConsumerImpl.h:58]:
```suggestion
                            Commands::SubscriptionMode = Commands::SubscriptionModeDurable,
```

File: lib/MultiTopicsConsumerImpl.h [hunk 8]
```diff
@@ -53,10 +54,15 @@ class MultiTopicsConsumerImpl : public ConsumerImplBase {
...
+                            boost::optional<MessageId> startMessageId = boost::none);
+
     MultiTopicsConsumerImpl(ClientImplPtr client, const std::vector<std::string>& topics,
                             const std::string& subscriptionName, TopicNamePtr topicName,
-                            const ConsumerConfiguration& conf, LookupServicePtr lookupServicePtr_);
+                            const ConsumerConfiguration& conf, LookupServicePtr lookupServicePtr_,
+                            const Commands::SubscriptionMode = Commands::SubscriptionModeDurable,
```
Comment (BewareMyPower) [inline lib/MultiTopicsConsumerImpl.h:64]:
```suggestion
                            Commands::SubscriptionMode = Commands::SubscriptionModeDurable,
```

**Review by BewareMyPower — APPROVED** [review]
Overall LGTM. Just need to remove some `const` qualifiers on value parameters.
